feat(guides): how to add delegation to a MUD project (#2127)
latticexyz/mud: docs/pages/guides/adding-delegation.mdx

import { CollapseCode } from "../../components/CollapseCode";

# Adding delegation

On this page you learn how to add [delegations](https://mud.dev/world/account-delegation) to enable addresses to act on behalf of other addresses.
This is important, for example, because typically every game move is a transaction, but you wouldn't want the player to constantly have to confirm every move on the wallet extension.

> The common pattern is to create a burner address and have the wallet user address authorize it to play on its behalf.

Reviewer: ```suggestion
The common pattern is to create a burner wallet and have the user's main wallet authorize it to play on its behalf.
```

## The sample application

> The application is located [in our monorepository](https://github.com/latticexyz/mud/tree/main/examples/multiple-accounts).

Reviewer: ```suggestion
The application is located [in the MUD monorepository](https://github.com/latticexyz/mud/tree/main/examples/multiple-accounts).
```

It creates multiple burner addresses, and keeps track onchain when they last submitted a transaction.

> There are the steps to run the initial application.

Reviewer: ```suggestion
Here are the steps to run the initial application:
```

> 1. Clone the monorepo, move the application out, and delete the monorepo again.

Reviewer: Why not just `cd mud/examples/multiple-accounts`?

Author: I just didn't think it was necessary to keep the entire repository. But you're right - I forgot disc space is a practically unlimited resource.

   ```sh copy
   git clone https://github.com/latticexyz/mud
   mv mud/examples/multiple-accounts .
   rm -rf mud
   cd multiple-accounts
   ```

> 1. Create `packages/contracts/.env`.

Reviewer: Should we commit this file with the example?

import { createPublicClient, createWalletClient, fallback, webSocket, http, ClientConfig, Hex } from "viem";
import { getNetworkConfig } from "./mud/getNetworkConfig";
import IWorldAbi from "contracts/out/IWorld.sol/IWorld.abi.json";

// The ABI we need
const ABI = [
  {
    inputs: [
      {
        internalType: "address",
        name: "delegatee",
        type: "address",
      },
      {
        internalType: "ResourceId",
        name: "systemId",
        type: "bytes32",
      },
      {
        internalType: "uint256",
        name: "numCalls",
        type: "uint256",
      },
    ],
    name: "initDelegation",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [],
    name: "newCall",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
];

// Create the system name constants

const name2hex = (name, length) => {
  let result = "";
  for (let i = 0; i < name.length; i++) result += name.charCodeAt(i).toString(16);

  if (length >= name.length) return result.padEnd(length * 2, "0");
  else return result.slice(0, length * 2);
};

const systemId = (namespace, systemName) =>

>   "0x" + name2hex("sy", 2) + name2hex(namespace, 14) + name2hex(systemName, 16);

Reviewer: we should use the `resourceToHex` util from `@latticexyz/common` here

const LASTCALL_SYSTEM_ID = systemId("LastCall", "LastCallSystem");
const SYSTEMBOUND_DELEGATION = systemId("", "systembound");

// A lot of code that would normally go in mud/getNetworkConfig.ts is part of the application here,
// because `getNetworkConfig.ts` assumes you use one wallet, and here we need several.
const networkConfig = await getNetworkConfig();

const clientOptions = {
  chain: networkConfig.chain,
  transport: transportObserver(fallback([webSocket(), http()])),
  pollingInterval: 1000,
} as const satisfies ClientConfig;

const publicClient = createPublicClient({
  ...clientOptions,
});

// Create a structure with two fields:
//   client - a wallet client that uses a random account
//   world - a world contract object that lets us issue newCall
const makeWorldContract = () => {
  const client = createWalletClient({
    ...clientOptions,
    account: createBurnerAccount(getBurnerPrivateKey(Math.random().toString())),
  });

  return {
    world: getContract({
      address: networkConfig.worldAddress as Hex,
      abi: IWorldAbi,
      publicClient: publicClient,
      walletClient: client,
    }),
    client,
  };
};

// Create five world contracts

> const worldContracts = [1, 2, 3, 4, 5].map((x) => makeWorldContract());

Reviewer: wait why do we need 5 world contracts?

Author: Because we have five different addresses (all burner wallets). We need to have multiple addresses to show delegation.

export const App = () => {
  const {
    network: { tables, useStore },
  } = useMUD();

  // Get all the calls from the records cache.
  const calls = useStore((state) => {
    const records = Object.values(state.getRecords(tables.LastCall));
    records.sort((a, b) => Number(a.value.callTime - b.value.callTime));
    return records;
  });

  // Convert timestamps to readable format
  const twoDigit = (str) => str.toString().padStart(2, "0");
  const timestamp2Str = (timestamp: number) => {
    const date = new Date(timestamp * 1000);
    return `${twoDigit(date.getHours())}:${twoDigit(date.getMinutes())}:${twoDigit(date.getSeconds())}`;
  };

  // Call newCall() on LastCall:LastCallSystem.
  const newCall = async (worldContract) => {

>     const tx = await worldContract.write.LastCall_LastCallSystem_newCall();

Reviewer: we need to update this with the latest version of MUD (no more system name in the function selector)
